repair: Avoid too many fragments in a single repair_row_on_wire

asias opened the pull request:
When repairing a partition with many rows, we can store many fragments in a repair_row_on_wire object which is sent as a rpc stream message.

This could cause reactor stalls when the rpc stream compression is turned on, because the compression compresses the whole message without any split and compression.

This patch solves the problem at the higher level by reducing the message size that is sent to the rpc stream.

Tests are added to make sure the message split works.

Fixes #24808

denesb commented:
@asias would it make sense to take the opportunity and also move to the v2 mutation fragment format? I think this is the only verb which uses it, no? Then we could solve both problems under a single feature flag.
If we plan to backport this (do we? -- I don't see backport labels), then we should think twice about linking the though -- although the v2 mutation fragments are now present in all releases, including enterprise ones.

avikivity commented on repair/row_level.cc:
```diff
@@ -767,6 +769,10 @@ class repair_meta {
     service::topology_guard _topology_guard;
+    // Max fragments size and number are allowed in a single repair_row_on_wire
+    // which is sent as a rpc stream message
+    static constexpr size_t _max_fragments_size = 1024 * 1024;
```
Any reason not to reduce it even more? Suggest 100k so there's some spare to reach 128k.

These are still streamed, yes? so there's no round-trip latency * the number of messages.

asias replied:
Each such message contains a partition key, I want to batch more fragments for such message. In most of the time, a reasonable total size of fragments allowed in a message would contain all the fragments needed to sync. If the number of fragments is huge, the partition key per such message is not that bad.  

Btw, we use std::list for the frozen_mutation_fragment stored in partition_key_and_mutation_fragments to avoid large allocations. 

```
class partition_key_and_mutation_fragments {
    partition_key _key;
    utils::chunked_vector<frozen_mutation_fragment> _mfs;
}
```

Will the 128K thing have impact at all?

avikivity replied:
We have `partition_key_and_mutation_fragments_v2 { optional<partition_key>, vector<fmf> };` Or just send a stream of mutation fragments.

1MB messages can cause stalls during allocation during compression. Let's reduce it to 240kB.

avikivity commented:
> @asias would it make sense to take the opportunity and also move to the v2 mutation fragment format? I think this is the only verb which uses it, no? Then we could solve both problems under a single feature flag. If we plan to backport this (do we? -- I don't see backport labels), then we should think twice about linking the though -- although the v2 mutation fragments are now present in all releases, including enterprise ones.

I suggest not to mix the two. One is a latency fix, the other refactoring.

I don't like growing our feature flag count, but they do allow us to be selective about backports.

denesb commented:
> > @asias would it make sense to take the opportunity and also move to the v2 mutation fragment format? I think this is the only verb which uses it, no? Then we could solve both problems under a single feature flag. If we plan to backport this (do we? -- I don't see backport labels), then we should think twice about linking the though -- although the v2 mutation fragments are now present in all releases, including enterprise ones.
> 
> I suggest not to mix the two. One is a latency fix, the other refactoring.
> 
> I don't like growing our feature flag count, but they do allow us to be selective about backports.

Yes, I also realized streaming is another user of mf v1 and piling that on top too would be way too much for this PR.

asias commented on repair/row_level.cc:
```diff
@@ -1443,11 +1449,28 @@ class repair_meta {
+            rlogger.info("Set inject_max_fragments_size {} for table={}.{}", max_fragments_size, ks, cf);
+        }
+        auto inject_max_fragments_nr = utils::get_local_injector().inject_parameter<uint32_t>("row_level_repair_max_fragments_nr");
+        if (inject_max_fragments_nr) {
```
We'd want not only the split but also how we split. The split on fragments nr allows us to split on a odd number of fragments to exercise the merge on the receiver side path. It is very important.

avikivity replied:
ok.

Deexie commented on repair/row_level.cc:
```diff
@@ -1456,15 +1479,26 @@ class repair_meta {
                 auto& row = rows.back();
-                if (last_dk_with_hash && dk_with_hash->dk.tri_compare(*_schema, last_dk_with_hash->dk) == 0) {
+                // Avoid too many fragments in a single repair_row_on_wire which is sent as a rpc stream message
+                bool no_split = msg_split_feature ? fragments_size < max_fragments_size && fragments_nr < max_fragments_nr : true;
```
What's the expected mf size? I wonder how much we can exceed the `max_fragments_size` and whether 
```
ragments_size + r.get_frozen_mutation().representation().size() < max_fragments_size
```
won't be safer

avikivity replied:
Yes. Most conservative logic is split if (new mf would cause overflow) && !there_are_zero_mutation_fragments.

Deexie commented on test/cluster/test_tablet_repair_scheduler.py:
```diff
@@ -336,3 +337,81 @@ async def check_filter():
+    await manager.server_start(servers[2].server_id)
+    await manager.server_start(servers[1].server_id)
+    await inject_error_on(manager, inject_error_name, servers, params={'value': inject_value})
+    await manager.api.tablet_repair(servers[0].ip_addr, ks, "test", token)
```
Don't we need to ensure that `servers[1]` (the one that receives the most data) isn't repair master?

asias commented:
Changes in v2:
- Reduce max fragment size to 240KB
- Do not send partition key if it is already sent in a previous repair_row_on_wire
- Change partition_key_and_mutation_fragments to use chunked_vector

mykaul commented:
```
 ld.lld: error: undefined symbol: to_repair_rows_list(std::__cxx11::list<partition_key_and_mutation_fragments, std::allocator<partition_key_and_mutation_fragments>>, seastar::lw_shared_ptr<schema const>, unsigned long, seastar::bool_class<repair_master_tag>, reader_permit, repair_hasher)
 >>> referenced by repair_test.cc:137 (./test/boost/repair_test.cc:137)
 >>>               build/dev/test/boost/repair_test.o:(seastar::noncopyable_function<void ()>::direct_vtable_for<seastar::futurize<std::invoke_result<repair_test::flush_repair_rows_on_wire_to_sstable::run_test_case() const::$_0>::type>::type seastar::async<repair_test::flush_repair_rows_on_wire_to_sstable::run_test_case() const::$_0>(seastar::thread_attributes, repair_test::flush_repair_rows_on_wire_to_sstable::run_test_case() const::$_0&&)::'lambda'()>::call(seastar::noncopyable_function<void ()> const*))
 >>> did you mean: to_repair_rows_list(std::__cxx11::list<partition_key_and_mutation_fragments, std::allocator<partition_key_and_mutation_fragments>>, seastar::lw_shared_ptr<schema const>, unsigned long, seastar::bool_class<repair_master_tag>, reader_permit, repair_hasher, bool)
 >>> defined in: build/dev/repair/row_level.o
 clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

asias commented:
> > ```
> >  ld.lld: error: undefined symbol: to_repair_rows_list(std::__cxx11::list<partition_key_and_mutation_fragments, std::allocator<partition_key_and_mutation_fragments>>, seastar::lw_shared_ptr<schema const>, unsigned long, seastar::bool_class<repair_master_tag>, reader_permit, repair_hasher)
> >  >>> referenced by repair_test.cc:137 (./test/boost/repair_test.cc:137)
> >  >>>               build/dev/test/boost/repair_test.o:(seastar::noncopyable_function<void ()>::direct_vtable_for<seastar::futurize<std::invoke_result<repair_test::flush_repair_rows_on_wire_to_sstable::run_test_case() const::$_0>::type>::type seastar::async<repair_test::flush_repair_rows_on_wire_to_sstable::run_test_case() const::$_0>(seastar::thread_attributes, repair_test::flush_repair_rows_on_wire_to_sstable::run_test_case() const::$_0&&)::'lambda'()>::call(seastar::noncopyable_function<void ()> const*))
> >  >>> did you mean: to_repair_rows_list(std::__cxx11::list<partition_key_and_mutation_fragments, std::allocator<partition_key_and_mutation_fragments>>, seastar::lw_shared_ptr<schema const>, unsigned long, seastar::bool_class<repair_master_tag>, reader_permit, repair_hasher, bool)
> >  >>> defined in: build/dev/repair/row_level.o
> >  clang++: error: linker command failed with exit code 1 (use -v to see invocation)
> > ```
> 
> This test is already fixed in the latest PR. It seems the CI used a older version.

I found the issue and pushed a new branch.

avikivity commented on repair/row_level.cc:
```diff
@@ -767,6 +774,10 @@ class repair_meta {
+    // Max fragments size and number are allowed in a single repair_row_on_wire
+    // which is sent as a rpc stream message
+    static constexpr size_t _max_fragments_size = 240 * 1024;
+    static constexpr size_t _max_fragments_nr = 32 * 1024;
```
But only if REPAIR_MSG_SPLIT is negotiated.

Note: REPAIR_MSG_SPLIT could appear in the middle of a repair. It's not correct for server parts of a client/server system to rely on it.

The correct approach is for the RPC that initiates the repair (the root of the RPC tree) to copy REPAIR_MSG_SPLIT into a request parameter, and everything else copies the flag from there, so everyone has a consistent view.

avikivity replied:
Please add documentation to idl/ to explain how it behaves differently. Usually we have [[version blah]] to indicate what changed (would be better to have [[feature blah]] now that I think about it).

I'd even prefer a new flag in repair_rows_on_wire than subtle changes in meaning of existing fields.

asias replied:
We still need the gossip feature bit otherwise the receiver with older version could not handle the split message.  I think it is better to add a new flag instead of using empty partition_key to ask the receiver to use the previous partition key. 

I will send a new version.

avikivity commented on repair/row_level.hh:
```diff
@@ -279,5 +279,5 @@ future<> repair_cf_range_row_level(repair::shard_repair_task_impl& shard_task,
 future<std::list<repair_row>> to_repair_rows_list(repair_rows_on_wire rows,
         schema_ptr s, uint64_t seed, repair_master is_master,
-        reader_permit permit, repair_hasher hasher);
+        reader_permit permit, repair_hasher hasher, bool msg_split_feature = false);
```
Better to avoid the default parameter, so we have to change all the call sites.

asias replied:
OK.  I will drop the flag in the new version though.

asias commented on utils/chunked_vector.hh:
```diff
@@ -656,7 +656,15 @@ chunked_vector<T, max_contiguous_allocation>::erase(const_iterator first, const_
+            pop_back();
+        }
+        shrink_to_fit();
+    }
```
OK. Done in new version.

asias commented on repair/row_level.cc:
```diff
@@ -1438,37 +1449,80 @@ class repair_meta {
             }
             co_await coroutine::maybe_yield();
         }
+        if (empty_partition_key) {
```
Because this version sends an empty partition (partition_key::make_empty()) as an indication that the receiver side should use the previous partition_key.  If we saw a empty partition that is read from the reader, we should should using it. But since it is confirmed that there will be no empty partition_key. I think we can throw in case one empty partition is emitted from the mutation reader. So we can simplify the logic.

asias commented on test/cluster/test_tablet_repair_scheduler.py:
```diff
@@ -336,3 +337,81 @@ async def check_filter():
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_tablet_repair_multiple_rows_merge_fragments_nr(manager: ManagerClient):
+    await run_tablet_repair_multiple_rows_merge(manager, "row_level_repair_max_fragments_nr", "9")
```
servers[1] will at most got two mutation fragments from the other two servers who will send its latest version. That's why an odd fragment_nr split will trigger the split that two messages share a same partition key which we want to test.

Deexie replied:
Ok, I understand, thank you!

asias commented:
Changes in v3: 

- Introduced use_last_partition_key flag instead of using empty partition_key to tell the receiver to use the previous partition key

- Introduced fragments_limiter helper
- Simplified split logic
- Added chunked_vector::resize_smaller for chunked_vector::erase

asias commented on repair/row_level.cc:
```diff
@@ -1442,30 +1449,70 @@ class repair_meta {
+            ++fragments_nr;
+        }
+        bool no_split(size_t size) {
+            return (fragments_size + size) < max_fragments_size && fragments_nr < max_fragments_nr;
```
I thought about this before and did not bother to change because the fragments_nr is mostly for testing purpose. I changed it in a new version.

asias commented on idl/repair.idl.hh:
```diff
@@ -22,7 +22,7 @@ class repair_hash {
 
 struct partition_key_and_mutation_fragments {
     partition_key get_key();
-    std::list<frozen_mutation_fragment> get_mutation_fragments();
```
Both commits change the same place. Swapping causes conflicts. I'd keep it as it is.

asias commented:
Changes in v5:
- Use fragments_nr + 1 for fragment nr check for consistency with size checking

asias commented:
> #### 🔴 CI State: FAILURE
> Mode	Stage	Status	Node
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18506/)	✅	spider9.cloudius-systems.com
> dev	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18506/)	✅	spider7.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18506/)	✅	spider7.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18506/)
> 🔹 test/boost/repair_test.cc
> 🔹 cluster/test_tablet_repair_scheduler	✅	spider7.cloudius-systems.com
> release	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18506/)	❌	spider2.cloudius-systems.com
> debug	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18506/)	❌	spider2.cloudius-systems.com
> #### Build Details:
> * Duration: 2 hr 45 min

CI  issue?

```
0:36:26  time="2025-07-24T05:36:25+03:00" level=warning msg="RunRoot is pointing to a path (/run/user/1000/containers) which is not writable. Most likely podman will fail."
10:36:26  Error: default OCI runtime "crun" not found: invalid argument
```

asias commented:
Changes in v6:

> > CI infra issue?
> > ```
> >  + podman login docker.io --username **** --password ****
> >  time="2025-07-24T09:31:03+03:00" level=warning msg="RunRoot is pointing to a path (/run/user/1000/containers) which is not writable. Most likely podman will fail."
> >  Error: default OCI runtime "crun" not found: invalid argument
> > script returned exit code 125
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @scylladb/releng ?
> 
> known issue , @Annamikhlin is working on it (already found a solution i assume) for now spider2 is out

OK. Thanks for the update.

yaronkaikov commented:
> > > CI infra issue?
> > > ```
> > >  + podman login docker.io --username **** --password ****
> > >  time="2025-07-24T09:31:03+03:00" level=warning msg="RunRoot is pointing to a path (/run/user/1000/containers) which is not writable. Most likely podman will fail."
> > >  Error: default OCI runtime "crun" not found: invalid argument
> > > script returned exit code 125
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > @scylladb/releng ?
> > 
> > 
> > known issue , @Annamikhlin is working on it (already found a solution i assume) for now spider2 is out
> 
> @mykaul @yaronkaikov Is the CI infra fix now. I still saw weird build issues in the other PR, e.g., [#24291 (comment)](https://github.com/scylladb/scylladb/pull/24291#issuecomment-3125455276).

yes it is

asias commented:
> > > > CI infra issue?
> > > > ```
> > > >  + podman login docker.io --username **** --password ****
> > > >  time="2025-07-24T09:31:03+03:00" level=warning msg="RunRoot is pointing to a path (/run/user/1000/containers) which is not writable. Most likely podman will fail."
> > > >  Error: default OCI runtime "crun" not found: invalid argument
> > > > script returned exit code 125
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > @scylladb/releng ?
> > > 
> > > 
> > > known issue , @Annamikhlin is working on it (already found a solution i assume) for now spider2 is out
> > 
> > 
> > @mykaul @yaronkaikov Is the CI infra fix now. I still saw weird build issues in the other PR, e.g., [#24291 (comment)](https://github.com/scylladb/scylladb/pull/24291#issuecomment-3125455276).
> 
> yes it is


See https://github.com/scylladb/scylladb/issues/25189#issuecomment-3130116769. Is the error here the same CI issue?


````
Regression
debug / test.py - debug / boost.memtable_test.cc.memtable_reader_after_tablet_migration.debug.3 (from pytest)

Failing for the past 1 build (Since [Failed#18570](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18570/) )
[Took 7.8 sec.](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18570/testReport/junit/boost.memtable_test/cc/debug___test_py___debug___memtable_reader_after_tablet_migration_debug_3/history)
[Add description](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18570/testReport/junit/boost.memtable_test/cc/debug___test_py___debug___memtable_reader_after_tablet_migration_debug_3/editDescription)
Error Message
working_dir: /jenkins/workspace/scylla-master/scylla-ci/scylla
Internal Error: calling /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/test/boost/combined_tests for test memtable_reader_after_tablet_migration failed (return_code=-6):
output file:/scylladir/testlog/x86_64/debug/memtable_reader_after_tablet_migration_stdout.3.log
log:
command to repeat:/jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/test/boost/combined_tests --report_level=no --output_format=XML --log_sink=/scylladir/testlog/x86_64/debug/memtable_reader_after_tablet_migration.3.log --catch_system_errors=no --color_output=false --run_test=memtable_test/memtable_reader_after_tablet_migration -- --overprovisioned --unsafe-bypass-fsync 1 --kernel-page-cache 1 --blocked-reactor-notify-ms 2000000 --collectd 0 --max-networking-io-control-blocks=100 --reactor-backend linux-aio -c2 -m2G
memtable_test.cc:-1: C++ failure
Stacktrace
working_dir: /jenkins/workspace/scylla-master/scylla-ci/scylla
Internal Error: calling /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/test/boost/combined_tests for test memtable_reader_after_tablet_migration failed (return_code=-6):
output file:/scylladir/testlog/x86_64/debug/memtable_reader_after_tablet_migration_stdout.3.log
log:
command to repeat:/jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/test/boost/combined_tests --report_level=no --output_format=XML --log_sink=/scylladir/testlog/x86_64/debug/memtable_reader_after_tablet_migration.3.log --catch_system_errors=no --color_output=false --run_test=memtable_test/memtable_reader_after_tablet_migration -- --overprovisioned --unsafe-bypass-fsync 1 --kernel-page-cache 1 --blocked-reactor-notify-ms 2000000 --collectd 0 --max-networking-io-control-blocks=100 --reactor-backend linux-aio -c2 -m2G
memtable_test.cc:-1: C++ failure
```

yaronkaikov commented:
> #24291 (comment)

This is a real issue, it's not related to CI infra. see https://github.com/scylladb/scylladb/issues/24792

denesb commented:
@asias please rebase, there are conflicts.

asias commented:
> @asias please rebase, there are conflicts.

Rebased.

asias commented:
> > > > CI infra issue?
> > > > ```
> > > >  + podman login docker.io --username **** --password ****
> > > >  time="2025-07-24T09:31:03+03:00" level=warning msg="RunRoot is pointing to a path (/run/user/1000/containers) which is not writable. Most likely podman will fail."
> > > >  Error: default OCI runtime "crun" not found: invalid argument
> > > > script returned exit code 125
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > @scylladb/releng ?
> > > 
> > > 
> > > known issue , @Annamikhlin is working on it (already found a solution i assume) for now spider2 is out
> > 
> > 
> > @mykaul @yaronkaikov Is the CI infra fix now. I still saw weird build issues in the other PR, e.g., [#24291 (comment)](https://github.com/scylladb/scylladb/pull/24291#issuecomment-3125455276).
> 
> yes it is

@yaronkaikov the release build is failing:

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18603/

denesb commented:
> Why have we decided not to backport this?

Our default is to not backport improvements.
I should have asked for justification for the backport decision (none), but forgot.

Is this issue a theoretical one or something we've seen in the field? The issue doesn't say.

bhalevy commented:
@Deexie / @denesb why isn't this backported?
It was hit in the field with 2025.1.2, for example.

asias commented:
> @Deexie / @denesb why isn't this backported? It was hit in the field with 2025.1.2, for example.

This is not a simple change to backport,  which even introduces a new gossip feature flag. I have concerns that the auto-backport-when-pr-is-merged-in-master is too aggressive.  It is not a correct issue, the worse case we saw so far is stall with the rpc compression.  I wish we do backport only after it stays in master longer.
